### PR TITLE
!!! FEATURE: AssetUsage as CatchUpHook

### LINKED ISSUE #4808: BUG: AssetUsageStrategy returns references to nodes of different contentstreams
**Current state:**
At the moment the AssetUsageStrategy fetches all usages of an asset from the AssetUsageFinder (over all content repositories) and groups them by node (`nodeaggregateid`, `origindimensionspacepointhash`). Which seems to be quite handy to figure out how many nodeaggregates and dimensions are using the asset.

But if you need to work with the referenced nodes, it doesn't respect the content streams where the node belongs to. Moreover, the group by chooses (by internal sorting) one of the contentstreams.

**Expected behavior**
In Neos <9.0 the AssetUsage returned **all usages of the `live` workspace** and the **usages of changed nodes in user workspaces** . Which seems to be the ideal for the output of the asset usage in the media module.

On the other hand we want to use the AssetUsage for content cache flushing, where we need exactly each asset usage per contentstream, to ensure all possible cache keys get flushed.

So we need to define what the expected behavior for Neos 9.0 is. And where to solve that. Do we need the same behavior as in Neos <9.0? Or do we specifiy a new behavior?

**Problems**

1. The AssetUsage Projection currently stores the usage for each workspace - independend of the change-state. So it's not possible to get just `live`-workspace and changed-node occurrences.
2. We have two different use cases, with different needs - but just one interface.

### LINKED ISSUE #5084: Adjust `AssetUsageProjection` to use workspace instead of content stream id
The AssetUsage needs to switch from contentStreamId to WorkspaceName from Projection until all depending DTOs.

* AssetUsageProjection
* AssetUsage
* AssetUsageFilter
* AssetUsageReference

With https://github.com/neos/neos-development-collection/pull/5002 we have access to the `WorkspaceName` in the projections instead of only the `ContentStreamId`.

That means we can adjust the write side + read side. And the chain down to the AssetUsageReference incl. AssetUsageFilter. To use workspace names.

```diff
final readonly class AssetUsageNodeAddress
{
    public function __construct(
-       public ContentStreamId $contentStreamId,
+       public WorkspaceName $workspaceName,
        public DimensionSpacePoint $dimensionSpacePoint,
        public NodeAggregateId $nodeAggregateId,
    ) {
    }
}
```

... as well as the `AssetUsage` class.

Should fix the following todos and problematic usages of the contentStreamId:
- https://github.com/neos/neos-development-collection/blob/67516e642e605f475bf07413b89be2e34dfc431f/Neos.Neos/Classes/AssetUsage/Service/AssetUsageSyncService.php#L58
- https://github.com/neos/neos-development-collection/blob/ec803ee9674f793426cdff944573c381135bdcc3/Neos.Neos/Classes/Service/ImageVariantGarbageCollector.php#L80
- https://github.com/neos/neos-development-collection/blob/9.0/Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php#L290

### PR DESCRIPTION
**Pull request by dlubitz**
This is an overhaul of the latest "asset usage" implementation which where using a dedicated projection to keep track of asset usages. As this was not handling all necessary events we tried to improve this, but got stuck due to the need of a lot of information of the graph, workspaces, etc (see https://github.com/neos/neos-development-collection/pull/5109). 

Therfore we decided to try a new approach to use a CatchUpHook instead, which enables us to access all needed information without re-implementing the whole graph logic. 

**General approach**
The package stores the usage of an asset per node and property in a dimension space point. It only stores the usage in the highest/base workspace (usually "live") as long as there are no changes in the same property in a lower/dependent workspace (e.g. user workspace). This allows to get easily an answer to the following questions:
1) Which nodes do I need to edit, to remove all usages for a safe asset removal (Neos Asset Usage/Media Browser)
2) Which cache entries do I need to flush on a change to an asset (this requires an additional traversal over all dependent workspaces).

**CatchUpHook**
The CatchUpHook is keeping track of all changes to node properties referencing an asset. On replay the CatchUpHook is not handling the events for performance reasons.

**Indexing Command**
The indexing command fills up the index initially. It can also be used to rebuild the index after projection replays or imports of events.

```
./flow assetusage:index --contentrepository default
```

Fixes: #4808 #5084

### CODE REVIEW

File: Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        if (!self::$enabled) {
+            // performance optimization: on full replay, we assume all caches to be flushed anyways
+            // - so we do not need to do it individually here.
+            return;
+        }
+
+        if (
+            $eventInstance instanceof EmbedsWorkspaceName
+            && $eventInstance instanceof EmbedsContentStreamId
+            && !$this->contentRepository->getWorkspaceFinder()->findOneByCurrentContentStreamId($eventInstance->getContentStreamId())?->workspaceName->equals($eventInstance->getWorkspaceName())
```
Comment (mhsdesign) [inline Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: i think we should prefer not using `findOneByCurrentContentStreamId` - also below.
```php
try {
    $contentGraph = $this->contentRepository->getContentGraph($eventInstance->getWorkspaceName());
    if (!$contentGraph->getContentStreamId()->equals($eventInstance->getContentStreamId()) {
        return;
    }
} catch (WorkspaceDoesNotExist) {
    return:
}
```

**Review by skurfuerst — APPROVED** [review]

File: Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php (hunk 1 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: try/catch for control flow is not great and costs performance

Comment (mhsdesign) [reply Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: but that is the bargain when we made `getContentGraph` throw instead of nullable ...

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -244,6 +244,7 @@ jobs:
           touch ${{ github.workspace }}/Data/DebugDatabaseDumps/keep
 
           ./flow package:list --loading-order
+          FLOW_CONTEXT=Testing/Behat ./flow doctrine:migrate --quiet
```
Comment (bwaidelich) [inline .github/workflows/build.yml:247]: (how) is that related?

File: Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php (hunk 1 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: @mhsdesign you're right, my comment wasn't really qualified.
And since we plan to remove the `WorkspaceFinder` with #5096 we should probably adjust this (we did a similar thing with #5270).

File: Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    public function onBeforeEvent(EventInterface $eventInstance, EventEnvelope $eventEnvelope): void
+    {
+        if (
+            $eventInstance instanceof EmbedsWorkspaceName
+            && $eventInstance instanceof EmbedsContentStreamId
+            // Safeguard for temporary content streams created during partial publish -> We want to skip these events, because their workspace doesn't match current contentstream.
+            && !$this->contentRepository->getWorkspaceFinder()->findOneByCurrentContentStreamId($eventInstance->getContentStreamId())?->workspaceName->equals($eventInstance->getWorkspaceName())
+        ) {
+            return;
+        }
```
Comment (bwaidelich) [inline Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: @dlubitz do you agree?
```suggestion
        if ($eventInstance instanceof EmbedsWorkspaceName && $eventInstance instanceof EmbedsContentStreamId) {
            // Safeguard for temporary content streams created during partial publish -> We want to skip these events, because their workspace doesn't match current content stream.
            try {
                $contentGraph = $this->contentRepository->getContentGraph($eventInstance->getWorkspaceName());
            } catch (WorkspaceDoesNotExist) {
                return:
            }
            if (!$contentGraph->getContentStreamId()->equals($eventInstance->getContentStreamId()) {
                return;
            }            
        }
```

BTW: Wasn't the plan to use some dedicated virtual workspace name for those temporary streams?

**Comment by bwaidelich** [comment]
Works like a charm, thanks for the hard work!!
Just a couple of comments, nothing severe!

File: .github/workflows/build.yml (hunk 2 above)
Comment (dlubitz) [reply .github/workflows/build.yml:247]: The behat tests are not working without this change, as the CatchUpHook expects the doctrine:migrate command to be executed. But it's not when we run the `Neos.ContentRepository.BehavioralTests`.

This was not an issue before, as these tests where only relying on the projection tables. But the new CatchUpHook now, gets executed also in these test cases.

https://github.com/neos/neos-development-collection/blob/9.0/.composer.json#L34-L41

File: Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php (hunk 3 above)
Comment (dlubitz) [reply Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: Yes this was the plan, but was not scope of this issue. This change just raised that.

Regarding the suggested change: Yeah, I'm fine with both as this will be removed/replaced anyway. Will change it later.

**Review by skurfuerst — APPROVED** [review]

File: Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php (hunk 3 above)
Comment (dlubitz) [reply Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: Done

File: Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php (hunk 1 above)
Comment (dlubitz) [reply Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: Done

Comment (mhsdesign) [reply Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php]: Thanks exactly my thoughts :D

File: .github/workflows/build.yml (hunk 2 above)
Comment (mhsdesign) [reply .github/workflows/build.yml:247]: That is our long standing: you have to do doctrine migrate sometimes problem otherwise it will not work problem that youll fix with the catchup thing as far as i know. see slack: https://neos-project.slack.com/archives/C04PYL8H3/p1727209924914109?thread_ts=1727117857.409749&cid=C04PYL8H3

**Comment by mhsdesign** [comment]
Post merge + 0.5 i just skimmed through most places thats why no +1 xD

File: .github/workflows/build.yml (hunk 2 above)
Comment (mhsdesign) [reply .github/workflows/build.yml:247]: Upps i was mistaken, this is just really similar to this problem: https://github.com/neos/neos-development-collection/pull/5005

In https://github.com/neos/neos-development-collection/issues/4878 i also discovered that disabling the catchup hooks will speed up the tests ... and as they are just run along without any assertions in the core tests we should consider doing this some point.

Comment (dlubitz) [reply .github/workflows/build.yml:247]: Yes, true. But we need to be able to enable it for the test cases where we test the catchup hook 🤷‍♂️
